### ISSUE TITLE
Source Generators: Post-build event

### ISSUE THREAD

**Issue by adampaquette** [dotnet/roslyn#66491]
# Background and Motivation

In most cases, source generators use attributes to influence the generation process but they contain many limitations. A much more powerful way is to use code. Here is an example from a library I wrote https://github.com/adampaquette/Typely that makes it easy to create value objects.

You define a class inheriting from `ITypelyConfiguration` which gives you access to a builder and all of its API methods. You can create extensions or functions to simplify your use cases, basically what you can think of.

```c#
public class TypesConfiguration : ITypelyConfiguration
{
    public void Configure(ITypelyBuilder builder)
    {
        builder.OfString().For("FirstName").NotEmpty();    
    }
}
```

This makes the generator "easy" to implement because it doesn't have to parse the `SyntaxTree` but executes the code with an injected `ITypelyBuilder` that builds an in-memory model of the desired types.

## Currently 
Because the source generator is triggered as the user types, the generator does not have access to a full blown assembly, so it generates an in-memory assembly from the `SyntaxTree` of the class, adds references, implicit using and executes the code. It dosen't have access to extensions or functions outside of the implemented `ITypelyConfiguration` class file. It cannot add ressources file (.resx) and compiles them easily...

## Expected
This is more of a discussion because I'm not sure exactly how this could be shaped but what I'm looking for is a post-build event that allows my generator to have access to all compiled assemblies and dependencies, to be able to dynamically execute the code to generate more code.

The tricky part is that some code, in the final assembly, needs the generator to have run. Maybe I could output a temp implementation with the comming Api https://github.com/dotnet/roslyn/issues/57589 for the post build to work.

I'm very open to suggestions / discussion.

# Proposed API
```c#
namespace Microsoft.CodeAnalysis
{
    public readonly partial struct IncrementalGeneratorInitializationContext
    {
+        public void RegisterPostBuildSourceOutput<TSource>(...);
     }
}
```

# Risks
Unsure of the implementation and impacts.
